airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/copy/s3/S3Copier.java

```diff
@@ -0,0 +1,210 @@
+                  SqlOperations sqlOperations)
+      throws IOException {
+    this.destSyncMode = destSyncMode;
+    this.schemaName = schema;
+    this.streamName = streamName;
+    this.s3Client = client;
+    this.db = db;
+    this.s3Config = s3Config;
+    this.nameTransformer = nameTransformer;
+    this.sqlOperations = sqlOperations;
+
+    this.s3StagingFile = String.join("/", stagingFolder, schemaName, streamName);
+
+    this.tmpTableName = nameTransformer.getTmpTableName(streamName);
+    // The stream transfer manager lets us greedily stream into S3. The native AWS SDK does not
+    // have support for streaming multipart uploads;
+    // The alternative is first writing the entire output to disk before loading into S3. This is not
+    // feasible with large tables.
+    // Data is chunked into parts. A part is sent off to a queue to be uploaded once it has reached it's
+    // configured part size.
+    // Memory consumption is queue capacity * part size = 10 * 10 = 100 MB at current configurations.
+    this.multipartUploadManager =
+        new StreamTransferManager(s3Config.getBucketName(), s3StagingFile, client)
+            .numUploadThreads(DEFAULT_UPLOAD_THREADS)
+            .queueCapacity(DEFAULT_QUEUE_CAPACITY)
+            .partSize(PART_SIZE_MB);
+    // We only need one output stream as we only have one input stream. This is reasonably performant.
+    // See the above comment.
+    this.outputStream = multipartUploadManager.getMultiPartOutputStreams().get(0);
+
+    var writer = new PrintWriter(outputStream, true, StandardCharsets.UTF_8);
+    this.csvPrinter = new CSVPrinter(writer, CSVFormat.DEFAULT);
+  }
+
+  @Override
+  public void write(UUID id, String jsonDataString, Timestamp emittedAt) throws Exception {
+    csvPrinter.printRecord(id, jsonDataString, emittedAt);
+  }
+
+  @Override
+  public String copyToTmpTableAndPrepMergeToFinalTable(boolean hasFailed) throws Exception {
+    if (hasFailed) {
+      multipartUploadManager.abort();
+      return "";
+    }
+    closeS3WriteStreamAndUpload();
+    createTmpTableAndCopyS3FileInto();
+    return mergeIntoDestTableIncrementalOrFullRefreshQuery();
+  }
+
+  @Override
+  public void removeFileAndDropTmpTable() throws Exception {
+    LOGGER.info("Begin cleaning s3 staging file {}.", s3StagingFile);
+    if (s3Client.doesObjectExist(s3Config.getBucketName(), s3StagingFile)) {
+      s3Client.deleteObject(s3Config.getBucketName(), s3StagingFile);
+    }
+    LOGGER.info("S3 staging file {} cleaned.", s3StagingFile);
+
+    LOGGER.info("Begin cleaning {} tmp table in destination.", tmpTableName);
+    sqlOperations.dropTableIfExists(db, schemaName, tmpTableName);
+    LOGGER.info("{} tmp table in destination cleaned.", tmpTableName);
+  }
+
+  private static String getFullS3Path(String s3BucketName, String s3StagingFile) {
+    return String.join("/", "s3:/", s3BucketName, s3StagingFile);
+  }
+
+  private void closeS3WriteStreamAndUpload() throws IOException {
```

Comment:
is this the equivalent to flushing the write queue? we're writing files into the tmp storage the whole time, but at the end we flush it?